source/agora/common/Channel.d

```diff
@@ -0,0 +1,535 @@
+/*******************************************************************************
+
+    This file contains the implementation of channels.
+    A channel is a communication class using which fiber can communicate
+    with each other.
+    Technically, a channel is a data transfer pipe where data can be passed
+    into or read from.
+    Hence one fiber can send data into a channel, while other fiber can read
+    that data from the same channel
+
+    Copyright:
+        Copyright (c) 2019 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.common.Channel;
+
+import std.container;
+import std.range;
+
+import core.sync.mutex;
+import core.thread;
+
+/// ditto
+public class Channel (T)
+{
+    /// closed
+    private bool closed;
+
+    /// lock
+    private Mutex mutex;
+
+    /// size of queue
+    private size_t qsize;
+
+    /// queue of data
+    private DList!T queue;
+
+    /// collection of send waiters
+    private DList!(SudoFiber!(T)) sendq;
+
+    /// collection of recv waiters
+    private DList!(SudoFiber!(T)) recvq;
+
+
+    /// Ctor
+    public this (size_t qsize = 0)
+    {
+        this.closed = false;
+        this.mutex = new Mutex;
+        this.qsize = qsize;
+    }
+
+    /***************************************************************************
+
+        Send data `elem`.
+        First, check the receiving waiter that is in the `recvq`.
+        If there are no targets there, add data to the `queue`.
+        If queue is full then stored waiter(fiber) to the `sendq`.
+
+        Params:
+            elem = value to send
+
+        Return:
+            true if the sending is successful, otherwise false
+
+    ***************************************************************************/
+
+    public bool send (T elem)
+    {
+        this.mutex.lock();
+
+        if (this.closed)
+        {
+            this.mutex.unlock();
+            return false;
+        }
+
+        if (this.recvq[].walkLength > 0)
+        {
+            SudoFiber!T sf = this.recvq.front;
+            this.recvq.removeFront();
+
+            this.mutex.unlock();
+
+            *(sf.elem_ptr) = elem;
+            if (sf.fiber !is null)
+                sf.fiber.call();
+            return true;
+        }
+
+        if (this.queue[].walkLength < this.qsize)
+        {
+            this.queue.insertBack(elem);
+            this.mutex.unlock();
+            return true;
+        }
+
+        Fiber fiber = Fiber.getThis();
+        if (fiber is null)
```

Comment:
Why is this creating a `Fiber` ? It should be possible to send a message between 2 threads without relying on Fibers.